File: quals/rev-thegoodgod/plus.c
#include <stdio.h>
#include <string.h>

#define myArray_size 29
#define value_size 57

// Values to camare against
const char res[] = {'\x40', '\x42', '\xb8', '\x4f', '\x92', '\xe5', '\x26', '\x33', '\xee', '\xa0', '\xc1', '\x97', '\xbc', '\x4f', '\x81', '\x43', '\x81', '\xe2', '\xdc', '\x2b', '\x92', '\xf9', '\x0f', '\x73', '\x96', '\x18', '\x2b', '\x33', '\xd0'};
// Sorted input 
char value[] = "33344445568EFLP___________aaadeeefhhiillmnnrsttttvwzzz{}";
char input[value_size];
char myArray[myArray_size];


void write_to(char base[], char c, size_t position) {
    size_t offset = position % 8;
    size_t first = position / 8 ;
    base[first] = base[first] ^ (c >> offset);
    if (offset == 0)
    {
        return ;
    }
    base[first+1] = base[first+1] ^ (c << (8 -offset));
}

/* 0 if correct, 1 otherwise */
int compare_contents(char input[], char value[], size_t val_size) {
    for (char* inp = input; *inp != '\0'; inp++)
    {
        for (size_t i = 0; i < val_size; i++)
        {
            if (value[i] == *inp)
            {
                value[i]='\0';
                break;
            }     
        }
    }
    
    char zeros[val_size];
    memset(zeros, 0, val_size);
    
    return memcmp(value, zeros, val_size);
}

int main(int argc, char const *argv[])
{
    
    printf("Your input: ");
    if (fgets(input, value_size, stdin)== NULL) {
        return 0;
    }
    input[value_size-1]='\0';    

    for (size_t i = 0; i < value_size; i++)
    {
        write_to(myArray, input[i], 4*i) ;
    }

    //if sorted(input) == value and myarray == value then "OK"
    if (! compare_contents(input, value, value_size) && ! memcmp(myArray, res, myArray_size))
    {
        printf("SUCCESS\n");
    } else {
        printf("FAIL\n");
    }
    
    return 0;
}

File: quals/rev-thegoodgod/plus.py
input = "Testing"
input = "EPFL{3z_disa55em8le_4_an_3z_r3v_with_4n_ez_fl46_at_th4t}"
step = 4
def write_to(base: list[int], c: str, position :int):
    offset = position % 8
    first = position // 8 
    c = ord(c)
    assert len(base) > first
    base[first] = (base[first] ^ (c >> offset))%256
    if (offset == 0) :
        return 
    assert len(base) > (first + 1) 
    base[first+1] = (base[first+1] ^ (c << (8 -offset)))%256

print("Actual size", len(input)*step/8)
myArray_size = int((len(input)*step/8 +1) // 1)
#print(myArray_size) 
l = [0]*myArray_size
for i in range(len(input)):
    write_to(l, input[i], step*i)
print("myArray_size = ", myArray_size)
print("value_size = ", len(input)+1)
#print(l)
print('{', end='')
print(*["'\\"+ f"{e:#0{4}x}'"[1:] for e in l], end='}\n', sep=', ' )
#print('{',",".join([f"'{e}'" for e in sorted(input)]),", '\\0'}")
print("".join(sorted(input)))
#print([hex(int(a)) for a in list(set("".join(sorted(input))))])

"""
    for (size_t i = 0; i < value_size; i++)
    {
        printf("%c\n", input[i]);
    }
    """